### ISSUE TITLE
Warnings not triggered when function goes through macro rename (with -isystem).

### ISSUE THREAD

**Issue by albestro** [llvm/llvm-project#53204]
Building our codebase we realized that the same kind of warning was triggered for CUSOLVER and not for CUBLAS.

After a bit of investigation, we saw that **the problem is related to functions, defined in header files included with the option `-isystem`, whose name gets "translated" by a macro define**.

For more details and a minimal reproducible example see https://github.com/eth-cscs/DLA-Future/pull/449#issuecomment-1012906903, but hoping to give a quick overview, next is reported the actual use-case with CUDA.

### Setup overview

CUBLAS, differently from CUSOLVER, provides an header file `cublas_v2.h` that looks like this

```cpp
#include "cublas_api.h"

#define cublasDgemm          cublasDgemm_v2
```

and then in `cublas_api.h` there is the definition of the function

```cpp
CUBLASAPI cublasStatus_t CUBLASWINAPI cublasDgemm_v2 (cublasHandle_t handle, //...
```

### The problem

Compiling with something like
```bash
clang++ ... -isystem /usr/local/cuda/include -Wall ...
```
using directly `cublasDgemm_v2` warnings are triggered, while using the macro `cublasDgemm` they are not.

On the opposite compiling with
```bash
clang++ ... -I /usr/local/cuda/include -Wall ...
```
warnings are always triggered (i.e. with macro or direct call).

I would say the expected result is to get the warnings also with the macro (this is also what happens with GCC).

**Comment by Artem-B** [comment]
> I would say the expected result is to get the warnings also with the macro

Can you provide a bit more details about the warnings you're concerned about, with specific warning messages?
The key is what those warnings are attributed to. 

My guess is that when you use `-isystem` for CUDA headers and the warning is triggered by a function argument.
When you use a macro, the macro expansion is attributed to the user code, but the macro expansion with the function call itself, and the corresponding warning gets attributed to the header. If that's a system header, the warning gets suppressed.

If you call the function directly, then the warning is attributed to the user code and you do end up seeing it.

AFAICT, this behavior is intentional. https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-in-system-headers

If you do want to enable warnings in the CUDA headers included with -isystem, you can use `--no-system-header-prefix=/usr/local/cuda/include`.

**Comment by albestro** [comment]
> > I would say the expected result is to get the warnings also with the macro
> 
> Can you provide a bit more details about the warnings you're concerned about, with specific warning messages? The key is what those warnings are attributed to.

In this https://github.com/eth-cscs/DLA-Future/pull/449#issuecomment-1012906903 you can see a more detailed report of the problem. Hope it helps.

> My guess is that when you use `-isystem` for CUDA headers and the warning is triggered by a function argument. When you use a macro, the macro expansion is attributed to the user code, but the macro expansion with the function call itself, and the corresponding warning gets attributed to the header. If that's a system header, the warning gets suppressed.
> 
> If you call the function directly, then the warning is attributed to the user code and you do end up seeing it.

It's a good point about what probably is going on under the hood. What I'm not sure about is if that's really the expected behavior.

In my mental model macros are not C++ first class citizens. The macro does not represent an entity in the language, it is just a way to manipulate text to produce code. In this case it just changes the name of the function, but semantically I would not say that the macro moves the calling point from user code to the header. The calling point is still in user-code, the name has just been changed by a macro.

I agree that this might be just my point of view. I don't pretend to be on the right side, just trying to clarify why I expected a different behavior 🙂

Just for reference, in the comment cited above, there is the comparison with another compiler. It does not mean that the other one is the correct behavior, but just that it sounds with what I was expecting.

> AFAICT, this behavior is intentional. https://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-in-system-headers

IIUC the doc talks about how "system-headers" works, i.e. they silence warnings. Reading that, I'm not able to say that the issue described above falls in the silenced cases. And I don't see any information about the assumption that macro "moves" the calling point.
Am I missing something?

**Comment by Artem-B** [comment]
> IIUC the doc talks about how "system-headers" works, i.e. they silence warnings. Reading that, I'm not able to say that the issue described above falls in the silenced cases. 

Whatever is included via `-isystem` makes it a 'system' header and will silence warnings there.

> And I don't see any information about the assumption that macro "moves" the calling point. Am I missing something?

That was a guess, based on observed behavior and the general idea of how clang deals with headers and warnings triggered by something in them. It may be wrong, too.

What is your goal? 
- To emit all the warnings in CUDA headers when they are included with `-isystem`. If that's the case, use `--no-system-header-prefix=...`.
- To emit all the warnings, in all the headers included with `-isystem`? That's unlikely to happen.
- To emit a warning specifically in case when something passed to a function in system header triggers a warning and the function call is expanded by a macro also defined in a system header? I'm not convinced that's something that needs fixing. It's an unfortunate combination of using `-isystem`, CUDA header redefining only a function name, so macro expansion does affect only *part* of the call site which leads to the confusion whether the warning should be attributed to user code or does expanding function name by a macro in a system header makes it belong to a header, even though the arguments all come from the user sources unchanged? I can't say either of those choices is perfect. I think it's up to the user to tell compiler explicitly that they want all warnings from particular headers. `--no-system-header-prefix=...` avoids the problem:

```shell
$ ../bin/clang++ -Wshorten-64-to-32 -isystem syshdr -c main.cc --no-system-header-prefix=syshdr -Wall
main.cc:11:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  foo(n);      // reference:   foo is translated to bar (current file)
  ~~~ ^
main.cc:12:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  fuu(n);      // problematic: fuu is translated to baz (header file)
  ~~~ ^
main.cc:13:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  baz(n);      // proof:       baz is a direct call to the function in header file
  ~~~ ^
3 warnings generated
```

The only thing that makes the current situation somewhat confusing is that clang does not elaborate that the function name has been expanded from a macro, so we only get info on the top-level calls. Well, this may be working as intended, too, as the argument itself is not expanded by the macro. If I use `#define fuu(x) baz(x)`, then the macro expansion does get reported as well:

```shell
main.cc:12:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  fuu(n);      // problematic: fuu is translated to baz (header file)
  ~~~~^~
syshdr/syshdr.h:4:20: note: expanded from macro 'fuu'
#define fuu(a) baz(a)
               ~~~ ^
```

**Comment by albestro** [comment]
> What is your goal?

My expectation is getting warnings for external libraries function calls included with `-isystem` flag. To make it clear, like `gcc` does (see report in https://github.com/eth-cscs/DLA-Future/pull/449#issuecomment-1012906903).

> - To emit a warning specifically in case when something passed to a function in system header triggers a warning and the function call is expanded by a macro also defined in a system header?

I would say yes.

> I'm not convinced that's something that needs fixing. It's an unfortunate combination of using `-isystem`, CUDA header redefining only a function name, so macro expansion does affect only _part_ of the call site which leads to the confusion whether the warning should be attributed to user code or does expanding function name by a macro in a system header makes it belong to a header, even though the arguments all come from the user sources unchanged? I can't say either of those choices is perfect. 

Ok. My main point is that macro are not expected to change this behavior (as I described here https://github.com/llvm/llvm-project/issues/53204#issuecomment-1016679534).

If you think is ok like this, feel free to close the issue. It's not a big deal. I thought it was a good thing to report a behavior that does not look correct.

**Comment by Artem-B** [comment]
Reproducer:

main.cc:
```c++
#include <cstddef>

#include "syshdr.h"

#define foo bar

int bar(int);

int main() {
  std::ptrdiff_t n = 13;
  foo(n);      // reference:   foo is translated to bar (current file)
  fuu(n);      // problematic: fuu is translated to baz (header file)
  fuua(n);     // problematic: fuua is translated to baz (header file)
  baz(n);      // proof:       baz is a direct call to the function in header file
}
```
syshdr/syshdr.h
```c++
// fake_include.h
#pragma once

#define fuu baz
#define fuua(a) baz(a)

int baz(int);
```

fuu/fuua produce no diags by default:
```shell
% ../bin/clang++ -Wshorten-64-to-32 -isystem syshdr -c main.cc
main.cc:11:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  foo(n);      // reference:   foo is translated to bar (current file)
  ~~~ ^
main.cc:14:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  baz(n);      // proof:       baz is a direct call to the function in header file
  ~~~ ^
2 warnings generated.
```

Diags are there is the header is no a system one:
```shell
% ../bin/clang++ -Wshorten-64-to-32 -isystem syshdr -c main.cc --no-system-header-prefix=syshdr
main.cc:11:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  foo(n);      // reference:   foo is translated to bar (current file)
  ~~~ ^
main.cc:12:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  fuu(n);      // problematic: fuu is translated to baz (header file)
  ~~~ ^
main.cc:13:8: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  fuua(n);     // problematic: fuua is translated to baz (header file)
  ~~~~~^~
syshdr/syshdr.h:5:21: note: expanded from macro 'fuua'
#define fuua(a) baz(a)
                ~~~ ^
main.cc:14:7: warning: implicit conversion loses integer precision: 'std::ptrdiff_t' (aka 'long') to 'int' [-Wshorten-64-to-32]
  baz(n);      // proof:       baz is a direct call to the function in header file
  ~~~ ^
4 warnings generated.
```
